modules/planning/scenarios/valet_parking/valet_parking_scenario.cc

```diff
@@ -17,21 +17,206 @@
 /**
  * @file
  **/
+
+#include <vector>
+
 #include "modules/planning/scenarios/valet_parking/valet_parking_scenario.h"
+#include "modules/planning/scenarios/valet_parking/stage_approaching_parking_spot.h"
+#include "modules/planning/scenarios/valet_parking/stage_parking.h"
 
 namespace apollo {
 namespace planning {
 namespace scenario {
 namespace valet_parking {
+
+using apollo::common::Status;
+using apollo::common::VehicleState;
+using apollo::common::math::Box2d;
+using apollo::common::math::Vec2d;
+using apollo::hdmap::HDMapUtil;
+using apollo::hdmap::LaneSegment;
+using apollo::hdmap::ParkingSpaceInfoConstPtr;
+using apollo::hdmap::Path;
+
+apollo::common::util::Factory<
+    ScenarioConfig::StageType, Stage,
+    Stage* (*)(const ScenarioConfig::StageConfig& stage_config)>
+    ValetParkingScenario::s_stage_factory_;
+
+void ValetParkingScenario::Init() {
+  if (init_) {
+    return;
+  }
+
+  Scenario::Init();
+
+  if (!GetScenarioConfig()) {
+    AERROR << "fail to get scenario specific config";
+    return;
+  }
+
+  hdmap_ = hdmap::HDMapUtil::BaseMapPtr();
+  CHECK_NOTNULL(hdmap_);
+}
+
+void ValetParkingScenario::RegisterStages() {
+  if (s_stage_factory_.Empty()) {
+    s_stage_factory_.Clear();
+  }
+  s_stage_factory_.Register(
+      ScenarioConfig::VALET_PARKING_APPROACHING_PARKING_SPOT,
+      [](const ScenarioConfig::StageConfig& config) -> Stage* {
+        return new StageApproachingParkingSpot(config);
+      });
+  s_stage_factory_.Register(
+      ScenarioConfig::VALET_PARKING_PARKING,
+      [](const ScenarioConfig::StageConfig& config) -> Stage* {
+        return new StageParking(config);
+      });
+}
+
 std::unique_ptr<Stage> ValetParkingScenario::CreateStage(
     const ScenarioConfig::StageConfig& stage_config) {
-  return nullptr;
+  if (s_stage_factory_.Empty()) {
+    RegisterStages();
+  }
+  auto ptr = s_stage_factory_.CreateObjectOrNull(stage_config.stage_type(),
+                                                 stage_config);
+  if (ptr) {
+    ptr->SetContext(&context_);
+  }
+  return ptr;
+}
+
+bool ValetParkingScenario::GetScenarioConfig() {
+  if (!config_.has_valet_parking_config()) {
+    AERROR << "miss scenario specific config";
+    return false;
+  }
+  context_.scenario_config.CopyFrom(config_.valet_parking_config());
+  return true;
 }
 
 bool ValetParkingScenario::IsTransferable(const Scenario& current_scenario,
                                           const Frame& frame) {
+  // TODO(all) Implement avaliable parking spot detection by preception results
```

Comment:
You can add a dummy function "IsParkingSpotAvailable"  and always returns true as of now.

Reply:
Will do that later, if we don't take available parking spot from routing message